File: backend/internal/api/handler/insight.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"insta360-insight/internal/model"
	"insta360-insight/internal/service"
)

// Insight 数据洞察 HTTP 处理器
type Insight struct {
	svc *service.InsightService
	ai  *service.AIService
}

func NewInsight(svc *service.InsightService, ai *service.AIService) *Insight {
	return &Insight{svc: svc, ai: ai}
}

// bindFilter 解析 querystring 中的筛选条件。
func bindFilter(c *gin.Context) model.Filter {
	return model.Filter{
		DateRange: c.QueryArray("date_range"),
		Regions:   c.QueryArray("regions"),
		Tracks:    c.QueryArray("tracks"),
		Platforms: c.QueryArray("platforms"),
		AgeBands:  c.QueryArray("age_bands"),
	}
}

func fail(c *gin.Context, err error) {
	c.JSON(http.StatusBadGateway, model.Fail(500, err.Error()))
}

// Kpi GET /api/kpi
func (h *Insight) Kpi(c *gin.Context) {
	data, err := h.svc.Kpi(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// ViewsTrend GET /api/views-trend
func (h *Insight) ViewsTrend(c *gin.Context) {
	data, err := h.svc.ViewsTrend(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// PlatformShare GET /api/platform-distribution
func (h *Insight) PlatformShare(c *gin.Context) {
	data, err := h.svc.PlatformShare(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// TrackPerformance GET /api/track-performance
func (h *Insight) TrackPerformance(c *gin.Context) {
	data, err := h.svc.TrackPerformance(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// Radar GET /api/explosive-radar
func (h *Insight) Radar(c *gin.Context) {
	data, err := h.svc.Radar(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// AudienceAge GET /api/audience-age
func (h *Insight) AudienceAge(c *gin.Context) {
	data, err := h.svc.AudienceAge(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// TopCreators GET /api/top-creators
func (h *Insight) TopCreators(c *gin.Context) {
	data, err := h.svc.TopCreators(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// Insights GET /api/insights —— AI 关键洞察
func (h *Insight) Insights(c *gin.Context) {
	data, err := h.ai.Generate(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

// FilterOptions GET /api/filters/options
func (h *Insight) FilterOptions(c *gin.Context) {
	data, err := h.svc.Options(c.Request.Context(), bindFilter(c))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, model.OK(data))
}

File: frontend/src/api/insight.js
// 数据洞察相关 API
// 全部走 /api 代理(前端 dev server 5173 -> 后端 8080)
import request from './request'

// querystring 序列化由 request.js 的自定义 paramsSerializer 处理:
// 数组参数展开为同名 key 重复(platforms=抖音&platforms=B站), 不带 [] 后缀,
// 与后端 gin c.QueryArray("platforms") 完美匹配。

const insight = {
  kpi: (q) => request.get('/kpi', { params: q }).then(r => r.data),
  viewsTrend: (q) => request.get('/views-trend', { params: q }).then(r => r.data),
  platformShare: (q) => request.get('/platform-distribution', { params: q }).then(r => r.data),
  tracks: (q) => request.get('/track-performance', { params: q }).then(r => r.data),
  radar: (q) => request.get('/explosive-radar', { params: q }).then(r => r.data),
  age: (q) => request.get('/audience-age', { params: q }).then(r => r.data),
  insights: (q) => request.get('/insights', { params: q }).then(r => r.data),
  topCreators: (q) => request.get('/top-creators', { params: q }).then(r => r.data),
  options: () => request.get('/filters/options').then(r => r.data)
}

export default insight
